hawk/app/models/report.rb

```diff
@@ -1,69 +1,225 @@
+      peinputs_raw, err, status = Util.capture3("crm", "history", :stdin_data => "source #{source}\npeinputs\n")
+      if status.exitstatus == 0
+        peinputs_raw.split(/\n/).each do |path|
+          next unless File.exists?(path)
+          v = peinput_version path
+          if v && v != pcmk_version
+            version = _("PE Input created by different Pacemaker version (%{version})" % { :version => v })
+          elsif v != pcmk_version
+            version = _("Pacemaker version not present in PE Input")
+          else
+            version = nil
+          end
+          peinputs.push(timestamp: File.mtime(path).iso8601,
+                        basename: File.basename(path, ".bz2"),
+                        filename: File.basename(path),
+                        path: path.sub("#{hb_report.path}/", ''),  # only use relative portion
+                        node: path.split(File::SEPARATOR)[-3],
+                        version: version)
+        end
+        # sort is going to be off for identical mtimes (stripped back to the second),
+        # so need secondary sort by filename
+      end
+
+      # add errors to output
+      errors = hb_report.err_filtered
+      errors.each do |err|
+        peinputs.push error: err
+      end
+    end
   end
 
-  def report_path
-    self.class.report_path
+  def peinput(path, basename, node)
+    basename = basename.gsub(/[^\w-]/, "")
+    node = node.gsub(/[^\w_-]/, "")
+    tname = "#{node}/pengine/#{basename}.bz2"
+    path = path.gsub("..", "") # tear out possible relative junk
+    archive.join(path).to_s
   end
 
-  class << self
-    def parse(file)
-      hash = Digest::SHA1.hexdigest(
-        file.basename(".tar.bz2").to_s
-      )
+  def peinput_version(path)
+    nvpair = `CIB_file=#{path} cibadmin -Q --xpath "/cib/configuration//crm_config//nvpair[@name='dc-version']" 2>/dev/null`
+    m = nvpair.match(/value="([^"]+)"/)
+    return nil unless m
+    m[1]
+  end
+
+  def transition_cmd(hb_report, path, cmd)
+    source = archive
+    source = hb_report.path if File.directory?(hb_report.path)
+    Util.capture3("crm", "history", stdin_data: "source #{source}\n#{cmd}\n")
+  end
+
+  def info(hb_report, path)
+    out, err, status = transition_cmd hb_report, path, "transition #{path} nograph"
+    info = out + err
+    info.strip!
+    info = _("No details available") if info.empty?
+    info.insert(0, _("Error:") + "\n") unless status.exitstatus == 0
+    info
+  end
+
+  def tags(hb_report, path)
+    out, err, status = transition_cmd hb_report, path, "transition tags #{path}"
+    out.split
+  end
+
+  def logs(hb_report, path)
+    out, err, status = transition_cmd hb_report, path, "transition log #{path}"
+    info = out + err
+    info.strip!
+    info = _("No details available") if info.empty?
+    info.insert(0, _("Error:") + "\n") unless status.exitstatus == 0
+    info
+  end
+
+  # Returns [success, data|error]
+  def graph(hb_report, path, format=:svg)
```

Comment:
Metrics/AbcSize: Assignment Branch Condition size for graph is too high. [30.1/15]<br>Style/SpaceAroundEqualsInParameterDefault: Surrounding space missing in default value assignment.